cmd/queue/main.go

```diff
@@ -322,17 +321,37 @@ func main() {
 		fmt.Sprintf(":%d", queue.RequestQueuePort),
 		queue.TimeToFirstByteTimeoutHandler(http.HandlerFunc(handler), time.Duration(revisionTimeoutSeconds)*time.Second, "request timeout"))
 
-	go server.ListenAndServe()
-	go setupAdminHandlers(adminServer)
+	errChan := make(chan error, 2)
+	// Runs a server created by creator and sends fatal errors to the errChan.
+	// Does not act on the ErrServerClosed error since that indicates we're
+	// already shutting everything down.
+	catchServerError := func(creator func() error) {
+		if err := creator(); err != nil && err != http.ErrServerClosed {
+			errChan <- err
+		}
+	}
+
+	setupAdminHandlers(adminServer)
+	go catchServerError(server.ListenAndServe)
+	go catchServerError(adminServer.ListenAndServe)
 
 	// Shutdown logic and signal handling
 	sigTermChan := make(chan os.Signal)
 	signal.Notify(sigTermChan, syscall.SIGTERM)
-	// Blocks until we actually receive a TERM signal.
-	<-sigTermChan
+
+	// Blocks until we actually receive a TERM signal or one of the servers
+	// exit unexpectedly. We fold both signals together because we only want
+	// to act on the first of those to reach here.
```

Comment:
I think we could replace most of this with https://godoc.org/golang.org/x/sync/errgroup#Group.Go.

Reply:
@greghaynes could you sketch how you'd envision that? The premise being: I'd like to keep the semantics of either signal or forceful abortion through failure and an error being thrown by the servers should not halt an ongoing graceful shutdown through a signal.

Reply:
@markusthoemmes heres what im thinking: https://github.com/knative/serving/compare/master...greghaynes:feature/queue-proxy-errgroup

Reply:
That does make sense indeed, especially because of the "return on first error" semantics that the errgroup has going on. Thanks for the tip.

Refactored the code a bit with errgroup.

Reply:
I opted for still catching the ErrServerClosed error though, since otherwise the flow becomes a little weird.

We'll shutdown the main server via the /quitquitquit handler, which will cause the error there to be returned and thus the wait to finish. The only reason the queue-proxy would be kept alive then is because there is still a connection to the adminServer (the quitquitquit connection) which keeps it open.

That should work in principle, I find it a lot more straightforward though, to actually wait for the SIGTERM to hit the container after the quitquiquit has finished and thus triggering the rest of the shutdown logic.